base/acme/src/org/dogtagpki/acme/validator/DNS01Validator.java

```diff
@@ -0,0 +1,158 @@
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.ResponseBuilder;
+import javax.ws.rs.core.UriInfo;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.lang.RandomStringUtils;
+import org.dogtagpki.acme.ACMEAuthorization;
+import org.dogtagpki.acme.ACMEChallenge;
+import org.dogtagpki.acme.ACMEError;
+import org.dogtagpki.acme.ACMEIdentifier;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class DNS01Validator extends ACMEValidator {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DNS01Validator.class);
+
+    public ACMEChallenge createChallenge(
+            UriInfo uriInfo,
+            String authzID,
+            String token) throws Exception {
+
+        // TODO: find better way to generate challenge ID
+
+        String challengeID = RandomStringUtils.randomAlphanumeric(10);
+        logger.info("Creating DNS challenge: " + challengeID);
+
+        URI challengeURI = uriInfo.getBaseUriBuilder().path("chall").path(challengeID).build();
+
+        ACMEChallenge challenge = new ACMEChallenge();
+        challenge.setID(challengeID);
+        challenge.setAuthzID(authzID);
+        challenge.setType("dns-01");
+        challenge.setURL(challengeURI);
+        challenge.setToken(token);
+
+        return challenge;
+    }
+
+    public String generateKeyAuthorization(String accountID, ACMEChallenge challenge) throws Exception {
+
+        // DNS-01 key authorization
+        // https://tools.ietf.org/html/rfc8555
+        // TODO: make it configurable
+
+        String keyAuthorization = challenge.getToken() + "." + accountID;
+        MessageDigest digest = MessageDigest.getInstance("SHA-256", "Mozilla-JSS");
+        byte[] hash = digest.digest(keyAuthorization.getBytes("UTF-8"));
+        return Base64.encodeBase64URLSafeString(hash);
+    }
+
+    public void validateChallenge(
+            ACMEAuthorization authorization,
+            ACMEChallenge challenge) throws Exception {
+
+        String accountID = authorization.getAccountID();
+        String keyAuthorization = generateKeyAuthorization(accountID, challenge);
+        logger.info("Key authorization: " + keyAuthorization);
+
+        ACMEIdentifier identifier = authorization.getIdentifier();
+        String hostname = identifier.getValue();
+        String recordName = "_acme-challenge." + hostname;
+
+        // TODO: make it configurable
+
+        int maxCount = 5;
+        int interval = 5;
+
+        int count = 1;
+        String value = null;
+        Exception exception = null;
+
+        while (true) {
+            try {
+                value = getTXTRecord(recordName);
+                break;
+            } catch (Exception e) {
+                logger.error(e.getMessage());
+                if (count >= maxCount) {
+                    exception = e;
+                    break;
+                }
+                Thread.sleep(interval * 1000);
```

Comment:
Not sure about LE Boulder, but with certbot you can take as much time as you need to configure the DNS and wait until it's properly propagated before you tell certbot to proceed with the verification step. So the number of retries and interval here is supposed to be used to handle network glitches during verification, not to wait for DNS configuration/propagation.

The current code is synchronous (and it works fine with certbot), but we probably should take a closer look at section 7.5.1 of RFC 8555. It might be possible to do it asynchronously since the client is supposed to poll the server to check for verification status.

We could use exponential back-off, but since it's only for network glitches I don't think it should grow very large. If the DNS record is still not ready after some reasonable time, we could assume that it's probably misconfigured and the user should try again with a new challenge.